### PR TITLE
Added tensorflow installation to repo

### PR DESCRIPTION
**Pull request by newtoncn**

### CODE REVIEW

**Comment by newtoncn** [comment]
[Add tensorflow installation to Jormungandr](https://trello.com/c/KU9k2ccz/23-add-tensorflow-installation-to-jormungandr)

**Comment by newtoncn** [comment]
I know it's pretty large, but I don't know if we'll get away with integrating tensorflow usefully without taking on some pretty heavy libraries

**Comment by garethellis0** [comment]
I'm fine with taking on heavy libraries if we need them, but checking them into our source control like this is probably not the best way to do it. The only files that should be maintained in our source control (outside of submodules) should be ones we own/maintain. If we want the tensorflow object detection api in here, it should be added as a git submodule. (Alternatively, see how Snowbots uses rosdep to install external dependencies: https://github.com/UBC-Snowbots/Snowflake/blob/master/.rosinstall).

**Comment by newtoncn** [comment]
@garethellis0 I think I've removed all unneeded files from this commit now

File: src/tensorflow_object_detector/data/labels/mscoco_label_map.pbtxt [hunk 1]
```diff
@@ -0,0 +1,400 @@
+item {
```
Comment (garethellis0) [inline src/tensorflow_object_detector/data/labels/mscoco_label_map.pbtxt]: This `data` folder doesn't seem like something we need to own? Unless we plan on modifying it? Is this part of the tensorflow models repo?

File: src/tensorflow_object_detector/package.xml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  <run_depend>sensor_msgs</run_depend>
+  <run_depend>std_msgs</run_depend>
+  <run_depend>vision_msgs</run_depend>
+
+  <!-- The export tag contains other, unspecified, tags -->
+  <export>
+    <!-- Other tools can request additional information be placed here -->
+
+  </export>
+</package>
```
Comment (garethellis0) [inline src/tensorflow_object_detector/package.xml]: We should remove commented out lines

File: src/tensorflow_object_detector/src/object_detection/utils/label_map_util.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    label_map_path: path to label_map.
+
+  Returns:
+    A dictionary mapping label names to id.
+  """
+  label_map = load_labelmap(label_map_path)
+  label_map_dict = {}
+  for item in label_map.item:
+    label_map_dict[item.name] = item.id
+  return label_map_dict
```
Comment (garethellis0) [inline src/tensorflow_object_detector/src/object_detection/utils/label_map_util.py]: Do we plan on modifying this file? Seems like it's from a library we could import?

File: src/tensorflow_object_detector/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+#############
+
+## Add gtest based cpp test target and link libraries
+# catkin_add_gtest(${PROJECT_NAME}-test test/test_tf_object_detect.cpp)
+# if(TARGET ${PROJECT_NAME}-test)
+#   target_link_libraries(${PROJECT_NAME}-test ${PROJECT_NAME})
+# endif()
+
+## Add folders to be run by python nosetests
+# catkin_add_nosetests(test)
```
[the remark this answers is not in the document]
Comment (newtoncn) [reply src/tensorflow_object_detector/CMakeLists.txt]: Removed unneeded comments

File: src/tensorflow_object_detector/src/object_detection/utils/label_map_util.py (hunk 3 above)
Comment (newtoncn) [reply src/tensorflow_object_detector/src/object_detection/utils/label_map_util.py]: No I don't plan on modifying this, this is just a utility to load in the pbtext files and create a map with them. It does mean that when we make our custom model we will just be able to change the labels in the labelmap and they will be properly loaded into the tensorflow sessiom. I plan on just modifying the detect_ros.py script when we make our own model

File: src/tensorflow_object_detector/package.xml (hunk 2 above)
Comment (newtoncn) [reply src/tensorflow_object_detector/package.xml]: Removed comments

**Comment by newtoncn** [comment]
The remaining scripts help convert label maps into one hot vectors and back,as well as creating the bounding box messages that will be published. I don't think we need to rewrite these scripts and I haven't found them in another library so I think we'll take them for now, if they cause problems I'll switch them out.

**Comment by garethellis0** [comment]
> The remaining scripts help convert label maps into one hot vectors and back,as well as creating the bounding box messages that will be published. I don't think we need to rewrite these scripts and I haven't found them in another library so I think we'll take them for now, if they cause problems I'll switch them out.

Oh, ok then. Are they just copied from stackoverflow or something (not that that's a bad thing, I've definitely done it before, just curious)?

**Comment by garethellis0** [comment]
Only thing still unresolved here is a few random data files in `src/tensorflow_object_detector/data`? Do we need these?

**Comment by garethellis0** [comment]
Makes sense. LGTM :+1: :tada:

**Comment by newtoncn** [comment]
@garethellis0 Thanks for the review!

**Pull request merged**